File: sampleValidation/python/jsonValidate.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

from __future__ import print_function
import sys
import os
import re
import json
import jsonschema as JSV
import uritools

# This is needed to assure open suports encoding parameter
if sys.version_info[0] > 2:
	# py3k
	pass
else:
	# py2
	import codecs
	import warnings
	def open(file, mode='r', buffering=-1, encoding=None, errors=None, newline=None, closefd=True, opener=None):
		if newline is not None:
			warnings.warn('newline is not supported in py2')
		if not closefd:
			warnings.warn('closefd is not supported in py2')
		if opener is not None:
			warnings.warn('opener is not supported in py2')
		return codecs.open(filename=file, mode=mode, encoding=encoding, errors=errors, buffering=buffering)

def getFKs(jsonSchema,jsonSchemaURI,prefix=""):
	FKs = []
	
	if isinstance(jsonSchema,dict):
		# First, this level's foreign keys
		isArray = False
		
		if 'items' in jsonSchema and isinstance(jsonSchema['items'],dict):
			jsonSchema = jsonSchema['items']
			isArray = True
			
			if prefix!='':
				prefix += '[]'
		
		if 'foreign_keys' in jsonSchema and isinstance(jsonSchema['foreign_keys'],(list,tuple)):
			for fk_def in jsonSchema['foreign_keys']:
				# Only valid declarations are taken into account
				if 'schema_id' in fk_def and 'members' in fk_def:
					ref_schema_id = fk_def['schema_id']
					members = fk_def['members']
					
					if isinstance(members,(list,tuple)):
						# Translating to absolute URI (in case it is relative)
						abs_ref_schema_id = uritools.urijoin(ref_schema_id,jsonSchemaURI)
						
						# Translating the paths
						components = tuple(map(lambda component: prefix + '.' + component  if component not in ['.','']  else prefix, members))
						
						FKs.append((abs_ref_schema_id,component))
		
		# Then, the foreign keys inside sublevels
		if 'properties' in jsonSchema and isinstance(jsonSchema['properties'],dict):
			if prefix != '':
				prefix += '.'
				p = jsonSchema['properties']
				for k,subSchema in p.items():
					push(FKs,getFKs(subSchema,jsonSchemaURI,prefix+k))
	
	return FKs


def loadJSONSchemas(p_schemaHash,*args):
	# Schema validation stats
	numDirOK = 0
	numDirFail = 0
	numFileOK = 0
	numFileIgnore = 0
	numFileFail = 0
	jsonSchemaFiles = list(args)
	for jsonSchemaFile in jsonSchemaFiles:
		if os.path.isdir(jsonSchemaFile):
			# It's a possible JSON Schema directory, not a JSON Schema file
			try:
				for relJsonSchemaFile in os.listdir(jsonSchemaFile):
					if relJsonSchemaFile[0]=='.':
						continue
					
					newJsonSchemaFile = os.path.join(jsonSchemaFile,relJsonSchemaFile)
					if os.path.isdir(newJsonSchemaFile) or '.json' in relJsonSchemaFile:
						jsonSchemaFiles.append(newJsonSchemaFile)
				numDirOK += 1
			except IOError as ioe:
				print("FATAL ERROR: Unable to open JSON schema directory {0}. Reason: {1}".format(jsonSchemaFile,ioe.strerror),file=sys.stderr)
				numDirFail += 1
		else:
			try:
				with open(jsonSchemaFile,mode="r",encoding="utf-8") as sHandle:
					print("* Loading schema {0}\n".format(jsonSchemaFile))
					
					jsonSchema = json.load(sHandle)
					
					valErrors = [ error  for error in JSV.validators.Draft4Validator(JSV.validators.Draft4Validator.META_SCHEMA).iter_errors(jsonSchema) ]
					if len(valErrors) > 0:
						print("\t- ERRORS:\n"+"\n".join(map(lambda se: "\t\tPath: {0} . Message: {1}".format(se.path,se.message) , valErrors)))
						numFileFail += 1
					else:
						# Getting the JSON Pointer object instance of the augmented schema
						# my $jsonSchemaP = $v->schema($jsonSchema)->schema;
						# This step is done, so we fetch a complete schema
						# $jsonSchema = $jsonSchemaP->data;
						if 'id' in jsonSchema:
							jsonSchemaURI = jsonSchema['id']
							if jsonSchemaURI in p_schemaHash:
								print("\tERROR: validated, but schema in {0} and schema in {1} have the same id\n".format(jsonSchemaFile,p_schemaHash[jsonSchemaURI][1]),file=sys.stderr)
								numFileFail += 1
							else:
								print("\t- Validated {0}\n".format(jsonSchemaURI))
								
								# Curating the primary key
								p_PK = None
								if 'primary_key' in jsonSchema:
									p_PK = jsonSchema['primary_key']
									if isinstance(p_PK,(list,tuple)):
										for key in p_PK:
											#if type(key) not in (int, long, float, bool, str):
											if type(key) not in (str,unicode):
												print("\tWARNING: primary key in {0} is not composed by strings defining its attributes. Ignoring it\n".format(jsonSchemaFile),file=sys.stderr)
												p_PK = None
												break
									else:
										p_PK = None
								
								# Gather foreign keys
								FKs = getFKs(jsonSchema,jsonSchemaURI)
								
								#use Data::Dumper;
								#
								#print STDERR Dumper(\@FKs),"\n";
								
								p_schemaHash[jsonSchemaURI] = (jsonSchema,jsonSchemaFile,p_PK,FKs)
								numFileOK += 1
						else:
							print("\tIGNORE: validated, but schema in {0} has no id attribute\n".format(jsonSchemaFile),file=sys.stderr)
							numFileIgnore += 1
			except IOError as ioe:
				print("FATAL ERROR: Unable to open schema file {0}. Reason: {1}\n".format(jsonSchemaFile,ioe.strerror),file=sys.stderr)
				numFileFail += 1
	
	print("\nSCHEMA VALIDATION STATS: loaded {0} schemas from {1} directories, ignored {2} schemas, failed {3} schemas and {4} directories\n".format(numFileOK,numDirOK,numFileIgnore,numFileFail,numDirFail))


jStepPat = re.compile(r"^([^\[]+)\[(0|[1-9][0-9]+)?\]$")
def materializeJPath(jsonDoc, jPath):
	objectives = [ jsonDoc ]
	jSteps = jPath.split('.') if jPath in ('.','') else (None,)
	for jStep in jSteps:
		newObjectives = []
		isArray = False
		arrayIndex = None
		if jStep is not None:
			jStepMatch = jStepPat.search(jStep)
			if jStepMatch is not None:
				isArray = True
				if jStepMatch.group(2) is not None:
					arrayIndex = int(jStepMatch.group(2))
				jStep = jStepMatch.group(1)
		for objective in objectives:
			if jStep is not None:
				if isinstance(objective,dict):
					value = objective[jStep]
				else:
					# Failing
					return None
			else:
				value = objective
			
			if isinstance(value,(list,tuple)):
				if arrayIndex is not None:
					newObjectives.append(value[arrayIndex])
				else:
					newObjectives.extend(value)
		
		objectives = newObjectives
	
	# Flattening it (we return a reference to a list of atomic values)
	for iobj, objective in enumerate(objectives):
		if not isinstance(objective,(int,long,str,unicode,float,bool)):
			objectives[iobj] = json.dumps(objective, sort_keys=True)
	
	return objectives


# It fetches the values from a JSON, based on the given paths to the members of the key
def getKeyValues(jsonDoc,p_members):
	return tuple(materializeJPath(jsonDoc,member) for member in p_members)


# It generates pk strings from a set of values
def genKeyStrings(keyTuple):
	pkStrings = []
	
	numPKcols = len(keyTuple)
	if numPKcols > 0:
		if isinstance(keyTuple[0],(list,tuple)):
			pkStrings = keyTuple[0]
			
			for curPKvalues in keyTuple[1:]:
				if isinstance(curPKvalues,(list, tuple)):
					newPKstrings = []
					
					for curPKvalue in curPKvalues:
						newPKstrings.extend(newPK+"\0"+curPKvalue  for newPK in newPKstrings)
					
					pkStrings = newPKstrings
				else:
					# If there is no found value, generate nothing
					pkStrings = []
					break
	
	return tuple(pkStrings)


def jsonValidate(p_schemaHash,*args):
	# A two level hash, in order to check primary key restrictions
	PKvals = dict()
	
	# JSON validation stats
	numDirOK = 0
	numDirFail = 0
	numFilePass1OK = 0
	numFilePass1Ignore = 0
	numFilePass1Fail = 0
	numFilePass2OK = 0
	numFilePass2Fail = 0
	
	# First pass, check against JSON schema, as well as primary keys unicity
	print("\nPASS 1: Schema validation and PK checks\n")
	iJsonFile = -1
	jsonFiles = list(args)
	for jsonFile in jsonFiles:
		iJsonFile += 1
		if os.path.isdir(jsonFile):
			# It's a possible JSON directory, not a JSON file
			try:
				for relJsonFile in os.listdir(jsonFile):
					# Skipping hidden files / directories
					if relJsonFile[0]=='.':
						continue
					
					newJsonFile = os.path.join(jsonFile,relJsonFile)
					if os.path.isdir(newJsonFile) or '.json' in relJsonFile:
						jsonFiles.append(newJsonFile)
				
				# Masking it for the pass 2 loop
				jsonFiles[iJsonFile] = None
				numDirOK += 1
			except IOError as ioe:
				print("FATAL ERROR: Unable to open JSON directory {0}. Reason: {1}\n".format(jsonFile,ioe.strerror),file=sys.stderr)
				numDirFail += 1
		else:
			try:
				with open(jsonFile,mode="r",encoding="utf-8") as jHandle:
					print("* Validating {0}".format(jsonFile))
					jsonDoc = json.load(jHandle)
					
					if '_schema' in jsonDoc:
						jsonSchemaId = jsonDoc['_schema']
						if jsonSchemaId in p_schemaHash:
							print("\t- Using {0} schema\n".format(jsonSchemaId))
							
							jsonSchema = p_schemaHash[jsonSchemaId][0]
							
							valErrors = [ error  for error in JSV.validators.Draft4Validator(jsonSchema).iter_errors(jsonDoc) ]
							
							if len(valErrors) > 0:
								print("\t- ERRORs:\n"+"\n".join(map(lambda se: "\t\tPath: {0} . Message: {1}".format(se.path,se.message) , valErrors)))
								numFilePass1Fail += 1
							else:
								# Does the schema contain a PK declaration?
								isValid = True
								p_PK_def = p_schemaHash[jsonSchemaId][2]
								if p_PK_def is not None:
									p_PK = None
									if jsonSchemaId in PKvals:
										p_PK = PKvals[jsonSchemaId]
									else:
										PKvals[jsonSchemaId] = p_PK = {}
									
									pkValues = getKeyValues(jsonDoc,p_PK_def)
									pkStrings = genKeyStrings(pkValues)
									for pkString in pkStrings:
										if pkString in p_PK:
											print("\t- PK ERROR: Duplicate PK in {0} and {1}\n".format(jsonFile,p_PK[pkString]),file=sys.stderr)
											isValid = False
										else:
											p_PK[pkString] = jsonFile
									
									# Masking it for the next loop if there was an error
									if not isValid:
										jsonFiles[iJsonFile] = None
										numFilePass1Fail += 1
										
								if isValid:
									print("\t- Validated!\n")
									numFilePass1OK += 1
							
						else:
							print("\t- Skipping schema validation (schema with URI {0} not found)\n".format(jsonSchemaId))
							# Masking it for the next loop
							jsonFiles[iJsonFile] = None
							numFilePass1Ignore += 1
					else:
						print("\t- Skipping schema validation (no one declared for {0})\n".format(jsonFile))
						# Masking it for the next loop
						jsonFiles[iJsonFile] = None
						numFilePass1Ignore += 1
					
					print("\n")
			except IOError as ioe:
				print("\t- ERROR: Unable to open file {0}. Reason: {1}\n".format(jsonFile,ioe.strerror),file=sys.stderr)
				numFilePass1Fail += 1
	
	#use Data::Dumper;
	#
	#print Dumper(\%PKvals),"\n";
	
	# Second pass, check foreign keys against gathered primary keys
	print("PASS 2: foreign keys checks\n")
	#use Data::Dumper;
	#print Dumper(@jsonFiles),"\n";
	for jsonFile in jsonFiles:
		if jsonFile is None:
			continue
		
		try:
			with open(jsonFile,mode="r",encoding="utf-8") as jHandle:
				print("* Checking FK on {0}".format(jsonFile))
				jsonDoc = json.load(jHandle)
				
				if '_schema' in jsonDoc:
					jsonSchemaId = jsonDoc['_schema']
					
					if jsonSchemaId in p_schemaHash:
						print("\t- Using {0} schema\n".format(jsonSchemaId))
						
						p_FKs = p_schemaHash[jsonSchemaId][3]
						
						isValid = True
						for p_FK_decl in p_FKs:
							pkSchemaId, p_FK_def = p_FK_decl
							
							fkValues = getKeyValues(jsonDoc,p_FK_def)
							#use Data::Dumper;
							#print Dumper(\@fkValues),"\n";
							
							fkStrings = genKeyStrings(fkValues)
							
							if len(fkStrings) > 0:
								if pkSckemaId in PKvals:
									p_PK = PKvals[pkSchemaId]
									for fkString in fkStrings:
										if fkString is not None:
											#print STDERR "DEBUG FK ",$fkString,"\n";
											if fkString not in p_PK:
												print("\t- FK ERROR: Missing FK to {0} in {1}\n".format(pkSchemaId,jsonFile),file=sys.stderr)
												isValid = False
												numFilePass2Fail += 1
												break
										#else:
										#	use Data::Dumper;
										#	print Dumper($p_FK_def),"\n";
								else:
									print("\t- FK ERROR: No available PKs ({0}) for {1}\n".format(pkSchemaId,jsonFile),file=sys.stderr)
									
									isValid = False
									numFilePass2Fail += 1
									break
						if isValid:
							print("\t- Validated!\n")
							numFilePass2OK += 1
					else:
						print("\t- ASSERTION ERROR: Skipping schema validation (schema with URI {0} not found)\n".format(jsonSchemaId))
						numFilePass2Fail += 1
				else:
					print("\t- ASSERTION ERROR: Skipping schema validation (no one declared for {0})\n".format(jsonFile))
					numFilePass2Fail += 1
		except IOError as ioe:
			print("\t- ERROR: Unable to open file {0}. Reason: {1}\n".format(jsonFile,ioe.strerror),file=sys.stderr)
			numFilePass2Fail += 1
	
	print("\nVALIDATION STATS:\n\t- directories ({0} OK, {1} failed)\n\t- PASS 1 ({2} OK, {3} ignored, {4} error)\n\t- PASS 2 ({5} OK, {6} error)\n".format(numDirOK,numDirFail,numFilePass1OK,numFilePass1Ignore,numFilePass1Fail,numFilePass2OK,numFilePass2Fail))


if len(sys.argv) > 1:
	jsonSchemaDir = sys.argv[1]
	
	jsonSchema = None
	
	schemaHash = {}
	loadJSONSchemas(schemaHash,jsonSchemaDir)
	
	if len(sys.argv) > 2:
		if len(schemaHash) == 0:
			print("FATAL ERROR: No schema was successfuly loaded. Exiting...\n",file=sys.stderr)
			sys.exit(1)
		
		args = tuple(sys.argv[2:])
		jsonValidate(schemaHash,*args)
else:
	print("Usage: {0} {{JSON schema}} {{JSON file}}*".format(sys.argv[0]),file=sys.stderr)
	sys.exit(1)

File: sampleValidation/perl/jsonValidate.pl
#!/usr/bin/perl

use warnings 'all';
use strict;

use FindBin qw();
use File::Spec qw();
use lib File::Spec->catfile($FindBin::Bin,'perllibs','lib','perl5');

use JSON::MaybeXS;
use JSON::Validator 0.95;

use URI;

use IO::Handle;
STDOUT->autoflush;
STDERR->autoflush;

sub getFKs($$;$);

sub getFKs($$;$) {
	my($jsonSchema,$jsonSchemaURI,$prefix) = @_;
	$prefix = ""  unless(defined($prefix));
	
	my @FKs = ();
	
	if(ref($jsonSchema) eq 'HASH') {
		# First, this level's foreign keys
		my $isArray = undef;
		
		if(exists($jsonSchema->{'items'}) && ref($jsonSchema->{'items'}) eq 'HASH') {
			$jsonSchema = $jsonSchema->{'items'};
			$isArray = 1;
			
			$prefix .= '[]'  if($prefix ne '');
		}
		
		if(exists($jsonSchema->{'foreign_keys'}) && ref($jsonSchema->{'foreign_keys'}) eq 'ARRAY') {
			foreach my $fk_def (@{$jsonSchema->{'foreign_keys'}}) {
				# Only valid declarations are taken into account
				if(exists($fk_def->{'schema_id'}) && exists($fk_def->{'members'})) {
					my($ref_schema_id,$members) = @{$fk_def}{'schema_id','members'};
					
					if(ref($members) eq 'ARRAY') {
						# Translating to absolute URI (in case it is relative)
						my $abs_ref_schema_id = URI->new_abs($ref_schema_id,$jsonSchemaURI);
						
						# Translating the paths
						my @components = map {  ($_ ne '.' && $_ ne '') ? $prefix . '.' . $_ : $prefix } @{$members};
						
						push(@FKs,[$abs_ref_schema_id,\@components]);
					}
				}
			}
		}
		
		# Then, the foreign keys inside sublevels
		if(exists($jsonSchema->{'properties'}) && ref($jsonSchema->{'properties'}) eq 'HASH') {
			$prefix .= "."  unless($prefix eq '');
			my $p = $jsonSchema->{'properties'};
			while(my($k,$subSchema) = each(%{$p})) {
				push(@FKs,getFKs($subSchema,$jsonSchemaURI,$prefix.$k));
			}
		}
	}
	
	return @FKs;
}

sub loadJSONSchemas(\%@) {
	my($p_schemaHash,@jsonSchemaFiles) = @_;
	my $p = JSON->new->convert_blessed;
	
	# Schema validation stats
	my $numDirOK = 0;
	my $numDirFail = 0;
	my $numFileOK = 0;
	my $numFileIgnore = 0;
	my $numFileFail = 0;
	foreach my $jsonSchemaFile (@jsonSchemaFiles) {
		if(-d $jsonSchemaFile) {
			# It's a possible JSON Schema directory, not a JSON Schema file
			if(opendir(my $JSD,$jsonSchemaFile)) {
				while(my $relJsonSchemaFile = readdir($JSD)) {
					# Skipping hidden files / directories
					next  if(substr($relJsonSchemaFile,0,1) eq '.');
					
					my $newJsonSchemaFile = File::Spec->catfile($jsonSchemaFile,$relJsonSchemaFile);
					push(@jsonSchemaFiles, $newJsonSchemaFile)  if(-d $newJsonSchemaFile || index($relJsonSchemaFile,'.json')!=-1);
				}
				closedir($JSD);
				$numDirOK++;
			} else {
				print STDERR "FATAL ERROR: Unable to open JSON schema directory $jsonSchemaFile. Reason: $!\n";
				$numDirFail++;
			}
		} else {
			if(open(my $S,'<:encoding(UTF-8)',$jsonSchemaFile)) {
				print "* Loading schema $jsonSchemaFile\n";
				local $/;
				my $jsonSchemaText = <$S>;
				close($S);
				
				my $jsonSchema = $p->decode($jsonSchemaText);
				my $v = JSON::Validator->new();
				my @valErrors = $v->schema(JSON::Validator::SPECIFICATION_URL)->validate($jsonSchema);
				if(scalar(@valErrors) > 0) {
					print "\t- ERRORS:\n".join("\n",map { "\t\tPath: ".$_->{'path'}.' . Message: '.$_->{'message'}} @valErrors)."\n";
					$numFileFail++;
				} else {
					# Getting the JSON Pointer object instance of the augmented schema
					my $jsonSchemaP = $v->schema($jsonSchema)->schema;
					# This step is done, so we fetch a complete schema
					$jsonSchema = $jsonSchemaP->data;
					if(exists($jsonSchema->{'id'})) {
						my $jsonSchemaURI = $jsonSchema->{'id'};
						if(exists($p_schemaHash->{$jsonSchemaURI})) {
							print STDERR "\tERROR: validated, but schema in $jsonSchemaFile and schema in ".$p_schemaHash->{$jsonSchemaURI}[1]." have the same id\n";
							$numFileFail++;
						} else {
							print "\t- Validated $jsonSchemaURI\n";
							
							# Curating the primary key
							my $p_PK = undef;
							if(exists($jsonSchema->{'primary_key'})) {
								$p_PK = $jsonSchema->{'primary_key'};
								if(ref($p_PK) eq 'ARRAY') {
									foreach my $key (@{$p_PK}) {
										if(ref(\$key) ne 'SCALAR') {
											print STDERR "\tWARNING: primary key in $jsonSchemaFile is not composed by strings defining its attributes. Ignoring it\n";
											$p_PK = undef;
											last;
										}
									}
								} else {
									$p_PK = undef;
								}
							}
							
							# Gather foreign keys 
							my @FKs = getFKs($jsonSchema,$jsonSchemaURI);
							
							#use Data::Dumper;
							#
							#print STDERR Dumper(\@FKs),"\n";
							
							$p_schemaHash->{$jsonSchemaURI} = [$jsonSchema,$jsonSchemaFile,$p_PK,\@FKs];
							$numFileOK++;
						}
					} else {
						print STDERR "\tIGNORE: validated, but schema in $jsonSchemaFile has no id attribute\n";
						$numFileIgnore++;
					}
				}
			} else {
				print STDERR "FATAL ERROR: Unable to open schema file $jsonSchemaFile. Reason: $!\n";
				$numFileFail++;
			}
		}
	}
	
	print "\nSCHEMA VALIDATION STATS: loaded $numFileOK schemas from $numDirOK directories, ignored $numFileIgnore schemas, failed $numFileFail schemas and $numDirFail directories\n";
}

sub materializeJPath($$) {
	my($jsonDoc,$jPath) = @_;
	
	my @objectives = ( $jsonDoc );
	my @jSteps = ($jPath eq '.' || $jPath eq '') ? ( undef ) : split(/\./,$jPath);
	foreach my $jStep (@jSteps) {
		my @newObjectives = ();
		my $isArray;
		my $arrayIndex;
		if(defined($jStep) && $jStep =~ /^([^\[]+)\[(0|[1-9][0-9]+)?\]$/) {
			$isArray = 1;
			$arrayIndex = $2 + 0  if(defined($2));
			$jStep = $1;
		}
		foreach my $objective ( @objectives ) {
			my $value;
			if(defined($jStep)) {
				if(ref($objective) eq 'HASH') {
					$value = $objective->{$jStep};
				} else {
					# Failing
					return undef;
				}
			} else {
				$value = $objective;
			}
			
			if(ref($value) eq 'ARRAY') {
				if(defined($arrayIndex)) {
					push(@newObjectives,$value->[$arrayIndex]);
				} else {
					push(@newObjectives,@{$value});
				}
			} else {
				push(@newObjectives,$value);
			}
		}
		
		@objectives = @newObjectives;
	}
	
	# Flattening it (we return a reference to a list of atomic values)
	foreach my $objective (@objectives) {
		if(ref($objective)) {
			$objective = JSON->new->convert_blessed->encode($objective);
		}
	}
	
	return \@objectives;
}

# It fetches the values from a JSON, based on the given paths to the members of the key
sub getKeyValues($\@) {
	my($jsonDoc,$p_members) = @_;
	
	return map { materializeJPath($jsonDoc,$_); } @{$p_members};
}

# It generates pk strings from a set of values
sub genKeyStrings(@) {
	my @pkStrings = ();
	
	my $numPKcols = scalar(@_);
	if($numPKcols > 0) {
		if(ref($_[0]) eq 'ARRAY') {
			@pkStrings = @{$_[0]};
			shift(@_);
			
			foreach my $curPKvalues (@_) {
				if(ref($curPKvalues) eq 'ARRAY') {
					my @newPKstrings = ();
					
					foreach my $curPKvalue (@{$curPKvalues}) {
						push(@newPKstrings,map { $_ . "\0" . $curPKvalue } @pkStrings);
					}
					
					@pkStrings = @newPKstrings;
				} else {
					# If there is no found value, generate nothing
					@pkStrings=();
					last;
				}
			}
		}
	}
	
	return @pkStrings;
}

sub jsonValidate(\%@) {
	my($p_schemaHash,@jsonFiles) = @_;
	my $p = JSON->new->convert_blessed;
	
	# A two level hash, in order to check primary key restrictions
	my %PKvals = ();
	
	# JSON validation stats
	my $numDirOK = 0;
	my $numDirFail = 0;
	my $numFilePass1OK = 0;
	my $numFilePass1Ignore = 0;
	my $numFilePass1Fail = 0;
	my $numFilePass2OK = 0;
	my $numFilePass2Fail = 0;
	
	# First pass, check against JSON schema, as well as primary keys unicity
	print "\nPASS 1: Schema validation and PK checks\n";
	foreach my $jsonFile (@jsonFiles) {
		if(-d $jsonFile) {
			# It's a possible JSON directory, not a JSON file
			if(opendir(my $JSD,$jsonFile)) {
				while(my $relJsonFile = readdir($JSD)) {
					# Skipping hidden files / directories
					next  if(substr($relJsonFile,0,1) eq '.');
					
					my $newJsonFile = File::Spec->catfile($jsonFile,$relJsonFile);
					push(@jsonFiles, $newJsonFile)  if(-d $newJsonFile || $relJsonFile =~ /\.json/);
				}
				closedir($JSD);
				
				# Masking it for the pass 2 loop
				$jsonFile = undef;
				$numDirOK++;
			} else {
				print STDERR "FATAL ERROR: Unable to open JSON directory $jsonFile. Reason: $!\n";
				$numDirFail++;
			}
		} elsif(open(my $J,'<:encoding(UTF-8)',$jsonFile)) {
			print "* Validating $jsonFile\n";
			local $/;
			my $jsonText = <$J>;
			close($J);
			
			my $jsonDoc = $p->decode($jsonText);
			
			if(exists($jsonDoc->{'_schema'})) {
				my $jsonSchemaId = $jsonDoc->{'_schema'};
				if(exists($p_schemaHash->{$jsonSchemaId})) {
					print "\t- Using $jsonSchemaId schema\n";
					
					my $jsonSchema = $p_schemaHash->{$jsonSchemaId}[0];
					
					my $v = JSON::Validator->new()->schema($jsonSchema);
					my @valErrors = $v->validate($jsonDoc);
					if(scalar(@valErrors) > 0) {
						print "\t- ERRORS:\n".join("\n",map { "\t\tPath: ".$_->{'path'}.' . Message: '.$_->{'message'}} @valErrors)."\n";
						
						# Masking it for the next loop
						$jsonFile = undef;
						$numFilePass1Fail++;
					} else {
						# Does the schema contain a PK declaration?
						my $isValid = 1;
						my $p_PK_def = $p_schemaHash->{$jsonSchemaId}[2];
						if(defined($p_PK_def)) {
							my $p_PK;
							if(exists($PKvals{$jsonSchemaId})) {
								$p_PK = $PKvals{$jsonSchemaId};
							} else {
								$PKvals{$jsonSchemaId} = $p_PK = {};
							}
							
							my @pkValues = getKeyValues($jsonDoc, @{$p_PK_def});
							my @pkStrings = genKeyStrings(@pkValues);
							foreach my $pkString (@pkStrings) {
								if(exists($p_PK->{$pkString})) {
									print STDERR "\t- PK ERROR: Duplicate PK in $jsonFile and ".$p_PK->{$pkString}."\n";
									$isValid = undef;
								} else {
									$p_PK->{$pkString} = $jsonFile;
								}
							}
							
							# Masking it for the next loop if there was an error
							unless($isValid) {
								$jsonFile = undef;
								$numFilePass1Fail++;
							}
						}
						
						if($isValid) {
							print "\t- Validated!\n";
							$numFilePass1OK++;
						}
					}
				} else {
					print "\t- Skipping schema validation (schema with URI ".$jsonSchemaId." not found)\n";
					# Masking it for the next loop
					$jsonFile = undef;
					$numFilePass1Ignore++;
				}
			} else {
				print "\t- Skipping schema validation (no one declared for $jsonFile)\n";
				# Masking it for the next loop
				$jsonFile = undef;
				$numFilePass1Ignore++;
			}
			print "\n";
		} else {
			print STDERR "\t- ERROR: Unable to open file $jsonFile. Reason: $!\n";
			# Masking it for the next loop
			$jsonFile = undef;
			$numFilePass1Fail++;
		}
	}
	
	#use Data::Dumper;
	#
	#print Dumper(\%PKvals),"\n";
	
	# Second pass, check foreign keys against gathered primary keys
	print "PASS 2: foreign keys checks\n";
	#use Data::Dumper;
	#print Dumper(@jsonFiles),"\n";
	foreach my $jsonFile (@jsonFiles) {
		next  unless(defined($jsonFile));
		
		if(open(my $J,'<:encoding(UTF-8)',$jsonFile)) {
			print "* Checking FK on $jsonFile\n";
			local $/;
			my $jsonText = <$J>;
			close($J);
			
			my $jsonDoc = $p->decode($jsonText);
			
			if(exists($jsonDoc->{'_schema'})) {
				my $jsonSchemaId = $jsonDoc->{'_schema'};
				if(exists($p_schemaHash->{$jsonSchemaId})) {
					print "\t- Using $jsonSchemaId schema\n";
					
					my $p_FKs = $p_schemaHash->{$jsonSchemaId}[3];
					
					my $isValid = 1;
					foreach my $p_FK_decl (@{$p_FKs}) {
						my($pkSchemaId,$p_FK_def) = @{$p_FK_decl};
						
						my @fkValues = getKeyValues($jsonDoc, @{$p_FK_def});
						#use Data::Dumper;
						#print Dumper(\@fkValues),"\n";
						
						my @fkStrings = genKeyStrings(@fkValues);
						
						if(scalar(@fkStrings) > 0) {
							if(exists($PKvals{$pkSchemaId})) {
								my $p_PK = $PKvals{$pkSchemaId};
								foreach my $fkString (@fkStrings) {
									if(defined($fkString)) {
										#print STDERR "DEBUG FK ",$fkString,"\n";
										unless(exists($p_PK->{$fkString})) {
											print STDERR "\t- FK ERROR: Missing FK to $pkSchemaId in $jsonFile\n";
											$isValid = undef;
											$numFilePass2Fail++;
											last;
										}
									#} else {
									#	use Data::Dumper;
									#	print Dumper($p_FK_def),"\n";
									}
								}
							} else {
								print STDERR "\t- FK ERROR: No available PKs ($pkSchemaId) for $jsonFile\n";
								
								$isValid = undef;
								$numFilePass2Fail++;
								last;
							}
						}
					}
					if($isValid) {
						print "\t- Validated!\n";
						$numFilePass2OK++;
					}
				} else {
					print "\t- ASSERTION ERROR: Skipping schema validation (schema with URI ".$jsonSchemaId." not found)\n";
					$numFilePass2Fail++;
				}
			} else {
				print STDERR "\t- ASSERTION ERROR: Skipping schema validation (no one declared for $jsonFile)\n";
				$numFilePass2Fail++;
			}
			print "\n";
		} else {
			print STDERR "\t- ERROR: Unable to open file $jsonFile. Reason: $!\n";
			$numFilePass2Fail++;
		}
	}
	
	print "\nVALIDATION STATS:\n\t- directories ($numDirOK OK, $numDirFail failed)\n\t- PASS 1 ($numFilePass1OK OK, $numFilePass1Ignore ignored, $numFilePass1Fail error)\n\t- PASS 2 ($numFilePass2OK OK, $numFilePass2Fail error)\n";
}

if(scalar(@ARGV) > 0) {
	my $jsonSchemaDir = shift(@ARGV);
	
	my $jsonSchema;
	
	my %schemaHash = ();
	loadJSONSchemas(%schemaHash,$jsonSchemaDir);
	
	if(scalar(@ARGV) > 0) {
		if(scalar(keys(%schemaHash))==0) {
			print STDERR "FATAL ERROR: No schema was successfuly loaded. Exiting...\n";
			exit 1;
		}
		
		jsonValidate(%schemaHash,@ARGV);
	}
} else {
	print STDERR "Usage: $0 {json_schema_directory} {json_file_or_directory}*\n";
	exit 1;
}
